From 78db825c9f4c9225a2db7f87c1b816f924f2f0cb Mon Sep 17 00:00:00 2001
From: Zhang Yanmin <yanmin.zhang@intel.com>
Date: Thu, 3 Mar 2016 10:07:32 +0800
Subject: [PATCH 1/5] pmc: dump wakeup status bit and rtc status

Sometimes, rtc in pmc might wake the system up when sleeping at s3.

The patch dumps wakeup status bit to check it and other checkers
if exist.

Change-Id: I70bac38ab065fd3b0c5534fd77308492ec7311be
Signed-off-by: Zhang Yanmin <yanmin.zhang@intel.com>
Signed-off-by: Li Youx <youx.li@intel.com>
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-23465
Reviewed-on: https://android.intel.com:443/479210
---
 arch/x86/kernel/pmc_atom.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/arch/x86/kernel/pmc_atom.c b/arch/x86/kernel/pmc_atom.c
index 946764b..6f0c5f6 100644
--- a/arch/x86/kernel/pmc_atom.c
+++ b/arch/x86/kernel/pmc_atom.c
@@ -25,6 +25,7 @@
 #include <linux/uaccess.h>
 #include <linux/suspend.h>
 
+#include <linux/syscore_ops.h>
 #include <asm/pmc_atom.h>
 
 #define	DRIVER_NAME	KBUILD_MODNAME
@@ -600,6 +601,28 @@ err:
 }
 #endif /* CONFIG_DEBUG_FS */
 
+#define	PMC_RTC_STS	(1 << 10)
+static void pmc_resume(void)
+{
+	u32 value;
+
+	if (!acpi_base_addr)
+		return;
+
+	value = inl(acpi_base_addr + 0);
+
+	value &= 0xFFFF;
+	pr_info("pmc: wakeup status[%x]\n", value);
+	if (value & PMC_RTC_STS)
+		pr_info("pmc: rtc might wake up system!\n");
+
+	return;
+}
+
+static struct syscore_ops pmc_syscore_ops = {
+	.resume = pmc_resume,
+};
+
 static int pmc_setup_dev(struct pci_dev *pdev)
 {
 	struct pmc_dev *pmc = &pmc_device;
@@ -673,6 +696,8 @@ static int __init pmc_atom_init(void)
 	struct pci_dev *pdev = NULL;
 	const struct pci_device_id *ent;
 
+	register_syscore_ops(&pmc_syscore_ops);
+
 	/* We look for our device - PCU PMC
 	 * we assume that there is max. one device.
 	 *
-- 
1.9.1

